src/common/libsubprocess/test/test_multi_echo.c

```diff
@@ -36,19 +45,57 @@ main (int argc, char *argv[])
         }
     }
 
+    if ((out + err) == 0) {
+        fprintf (stderr, "must specify a way to output");
+        exit (1);
+    }
+
     if (optind != argc) {
+        int maxcount = count + 1; /* for newline */
+
         while (optind < argc) {
-            int i;
-            for (i = 0; i < count; i++) {
-                printf ("%s", argv[optind]);
-                fflush (stdout);
+            int ret, outcount = 0, errcount = 0;
+
+            /* some tests can flood / hang fprintf calls, so check
+             * that we can output */
+            while ((out && outcount < maxcount)
+                   || (err && errcount < maxcount)) {
+                struct pollfd pfds[2];
+
+                pfds[0].fd = STDOUT_FILENO;
+                pfds[0].events = (out && outcount < maxcount) ? POLLOUT: 0;
+                pfds[0].revents = 0;
+
+                pfds[1].fd = STDERR_FILENO;
+                pfds[1].events = (err && errcount < maxcount) ? POLLOUT: 0;
+                pfds[1].revents = 0;
+
+                if ((ret = poll (pfds, 2, -1)) < 0) {
+                    perror ("poll");
+                    exit (1);
+                }
+                if (out && pfds[0].revents & POLLOUT) {
+                    if (outcount == count)
+                        fprintf (stdout, "\n");
+                    else
+                        fprintf (stdout, "%s", argv[optind]);
+                    fflush (stdout);
+                    outcount++;
+                }
+                if (err && pfds[1].revents & POLLOUT) {
+                    if (errcount == count)
+                        fprintf (stderr, "\n");
+                    else
+                        fprintf (stderr, "%s", argv[optind]);
+                    fflush (stderr);
+                    errcount++;
+                }
             }
             optind++;
         }
-        printf ("\n");
-        fflush (stdout);
     }
 
     close (STDOUT_FILENO);
```

Comment:
stdio atexit handler will close these after flushing I/O, so shouldn't be necessary here.